File: realm/realm-library/src/syncIntegrationTest/java/io/realm/SSLConfigurationTests.java
/*
 * Copyright 2017 Realm Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.realm;

import android.os.SystemClock;
import android.support.test.runner.AndroidJUnit4;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.Timeout;
import org.junit.runner.RunWith;

import java.util.UUID;
import java.util.concurrent.TimeUnit;

import io.realm.entities.StringOnly;
import io.realm.exceptions.RealmFileException;
import io.realm.log.LogLevel;
import io.realm.log.RealmLog;
import io.realm.objectserver.utils.Constants;
import io.realm.rule.RunTestInLooperThread;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

@RunWith(AndroidJUnit4.class)
public class SSLConfigurationTests extends StandardIntegrationTest {

    // TODO: All tests in this class are currently marked @RunTestInLooperThread,
    // this is strictly not necessary, but currently needed to avoid other issues with setting
    // up tests.

    @Rule
    public Timeout globalTimeout = Timeout.seconds(120);

    @Test
    @RunTestInLooperThread
    public void trustedRootCA() throws InterruptedException {
        String username = UUID.randomUUID().toString();
        String password = "password";
        SyncUser user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);

        // 1. Copy a valid Realm to the server
        //noinspection unchecked
        final SyncConfiguration syncConfig = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM)
                .fullSynchronization()
                .schema(StringOnly.class)
                .build();
        Realm realm = Realm.getInstance(syncConfig);

        realm.beginTransaction();
        realm.createObject(StringOnly.class).setChars("Foo");
        realm.commitTransaction();

        // make sure the changes gets to the server
        SyncManager.getSession(syncConfig).uploadAllLocalChanges();
        realm.close();
        user.logOut();

        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
        // download the uploaded changes.
        user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password), Constants.AUTH_URL);
        //noinspection unchecked
        SyncConfiguration syncConfigSSL = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM_SECURE)
                .fullSynchronization()
                .name("useSsl")
                .schema(StringOnly.class)
                .waitForInitialRemoteData()
                .trustedRootCA("trusted_ca.pem")
                .build();
        realm = Realm.getInstance(syncConfigSSL);

        RealmResults<StringOnly> all = realm.where(StringOnly.class).findAll();
        try {
            assertEquals(1, all.size());
            assertEquals("Foo", all.get(0).getChars());
        } finally {
            realm.close();
        }
        looperThread.testComplete();
    }

    @Test
    @RunTestInLooperThread
    public void withoutSSLVerification() throws InterruptedException {
        String username = UUID.randomUUID().toString();
        String password = "password";
        SyncUser user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);

        // 1. Copy a valid Realm to the server
        //noinspection unchecked
        final SyncConfiguration syncConfig = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM)
                .fullSynchronization()
                .schema(StringOnly.class)
                .build();
        Realm realm = Realm.getInstance(syncConfig);

        realm.beginTransaction();
        realm.createObject(StringOnly.class).setChars("Foo");
        realm.commitTransaction();

        // make sure the changes gets to the server
        SyncManager.getSession(syncConfig).uploadAllLocalChanges();
        realm.close();
        user.logOut();

        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
        // download the uploaded changes.
        user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password), Constants.AUTH_URL);
        //noinspection unchecked
        SyncConfiguration syncConfigSSL = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM_SECURE)
                .fullSynchronization()
                .name("useSsl")
                .schema(StringOnly.class)
                .waitForInitialRemoteData()
                .disableSSLVerification()
                .build();
        realm = Realm.getInstance(syncConfigSSL);

        RealmResults<StringOnly> all = realm.where(StringOnly.class).findAll();
        try {
            assertEquals(1, all.size());
            assertEquals("Foo", all.get(0).getChars());
        } finally {
            realm.close();
        }
        looperThread.testComplete();
    }

    @Test
    @RunTestInLooperThread
    public void trustedRootCA_syncShouldFailWithoutTrustedCA() throws InterruptedException {
        String username = UUID.randomUUID().toString();
        String password = "password";
        SyncUser user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);

        // 1. Copy a valid Realm to the server
        //noinspection unchecked
        final SyncConfiguration syncConfig = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM)
                .schema(StringOnly.class)
                .build();
        Realm realm = Realm.getInstance(syncConfig);

        realm.beginTransaction();
        realm.createObject(StringOnly.class).setChars("Foo");
        realm.commitTransaction();

        // make sure the changes gets to the server
        SyncManager.getSession(syncConfig).uploadAllLocalChanges();
        realm.close();
        user.logOut();

        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
        // download the uploaded changes.
        user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password), Constants.AUTH_URL);
        //noinspection unchecked
        SyncConfiguration syncConfigSSL = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM_SECURE)
                .name("useSsl")
                .schema(StringOnly.class)
                .trustedRootCA("untrusted_ca.pem")
                .build();
        // waitForInitialRemoteData will throw an Internal error (125): Operation Canceled
        SystemClock.sleep(TimeUnit.SECONDS.toMillis(2));
        realm = Realm.getInstance(syncConfigSSL);
        try {
            assertTrue(realm.isEmpty());
        } finally {
            realm.close();
        }
        looperThread.testComplete();
    }

    @Test
    @RunTestInLooperThread
    public void combining_trustedRootCA_and_withoutSSLVerification_willThrow() {
        String username = UUID.randomUUID().toString();
        String password = "password";
        SyncUser user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);

        TestHelper.TestLogger testLogger = new TestHelper.TestLogger();
        int originalLevel = RealmLog.getLevel();
        RealmLog.add(testLogger);
        RealmLog.setLevel(LogLevel.WARN);

        //noinspection unchecked
        configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM_SECURE)
                .name("useSsl")
                .schema(StringOnly.class)
                .trustedRootCA("trusted_ca.pem")
                .disableSSLVerification()
                .build();

        assertEquals("SSL Verification is disabled, the provided server certificate will not be used.",
                testLogger.message);
        RealmLog.remove(testLogger);
        RealmLog.setLevel(originalLevel);
        looperThread.testComplete();
    }

    @Test
    @RunTestInLooperThread
    public void trustedRootCA_notExisting_certificate_willThrow() {
        String username = UUID.randomUUID().toString();
        String password = "password";
        SyncUser user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);
        //noinspection unchecked
        SyncConfiguration syncConfig = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM_SECURE)
                .schema(StringOnly.class)
                .trustedRootCA("none_existing_file.pem")
                .build();

        try {
            Realm.getInstance(syncConfig);
            fail();
        } catch (RealmFileException ignored) {
        }
        looperThread.testComplete();
    }

    @Test
    @RunTestInLooperThread
    public void combiningTrustedRootCA_and_disableSSLVerification() throws InterruptedException {
        String username = UUID.randomUUID().toString();
        String password = "password";
        SyncUser user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);

        // 1. Copy a valid Realm to the server using ssl_verify_path option
        //noinspection unchecked
        final SyncConfiguration syncConfigWithCertificate = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM_SECURE)
                .fullSynchronization()
                .schema(StringOnly.class)
                .trustedRootCA("trusted_ca.pem")
                .build();
        Realm realm = Realm.getInstance(syncConfigWithCertificate);

        realm.beginTransaction();
        realm.createObject(StringOnly.class).setChars("Foo");
        realm.commitTransaction();

        // make sure the changes gets to the server
        SyncManager.getSession(syncConfigWithCertificate).uploadAllLocalChanges();
        realm.close();
        user.logOut();

        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
        // download the uploaded changes.
        user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password), Constants.AUTH_URL);
        //noinspection unchecked
        SyncConfiguration syncConfigDisableSSL = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM_SECURE)
                .fullSynchronization()
                .name("useSsl")
                .schema(StringOnly.class)
                .waitForInitialRemoteData()
                .disableSSLVerification()
                .build();
        realm = Realm.getInstance(syncConfigDisableSSL);

        RealmResults<StringOnly> all = realm.where(StringOnly.class).findAll();
        try {
            assertEquals(1, all.size());
            assertEquals("Foo", all.get(0).getChars());
        } finally {
            realm.close();
        }
        looperThread.testComplete();
    }

    // IMPORTANT: Following test assume the root certificate is installed on the test device
    //            certificate is located in <realm-java>/tools/sync_test_server/keys/android_test_certificate.crt
    //            adb push <realm-java>/tools/sync_test_server/keys/android_test_certificate.crt /sdcard/
    //            then import the certificate from the device (Settings/Security/Install from storage)
    @Test
    @RunTestInLooperThread
    public void sslVerifyCallback_isUsed() throws InterruptedException {
        String username = UUID.randomUUID().toString();
        String password = "password";
        SyncUser user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password, true), Constants.AUTH_URL);

        // 1. Copy a valid Realm to the server using ssl_verify_path option
        //noinspection unchecked
        final SyncConfiguration syncConfig = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM)
                .fullSynchronization()
                .schema(StringOnly.class)
                .build();
        Realm realm = Realm.getInstance(syncConfig);

        realm.beginTransaction();
        realm.createObject(StringOnly.class).setChars("Foo");
        realm.commitTransaction();

        // make sure the changes gets to the server
        SyncManager.getSession(syncConfig).uploadAllLocalChanges();
        realm.close();
        user.logOut();

        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
        // download the uploaded changes.
        user = SyncUser.logIn(SyncCredentials.usernamePassword(username, password), Constants.AUTH_URL);
        //noinspection unchecked
        SyncConfiguration syncConfigSecure = configurationFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM_SECURE)
                .name("useSsl")
                .fullSynchronization()
                .schema(StringOnly.class)
                .waitForInitialRemoteData()
                .build();
        realm = Realm.getInstance(syncConfigSecure);

        RealmResults<StringOnly> all = realm.where(StringOnly.class).findAll();
        try {
            assertEquals(1, all.size());
            assertEquals("Foo", all.get(0).getChars());
        } finally {
            realm.close();
        }
        looperThread.testComplete();
    }
}

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/SSLConfigurationTests.kt
/*
 * Copyright 2017 Realm Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.realm.mongodb.sync

import android.os.SystemClock
import androidx.test.ext.junit.runners.AndroidJUnit4
import io.realm.Realm
import io.realm.TestApp
import io.realm.TestHelper
import io.realm.TestHelper.TestLogger
import io.realm.TestSyncConfigurationFactory
import io.realm.entities.DefaultSyncSchema
import io.realm.entities.SyncDog
import io.realm.exceptions.RealmFileException
import io.realm.kotlin.syncSession
import io.realm.log.LogLevel
import io.realm.log.RealmLog
import io.realm.mongodb.App
import io.realm.mongodb.AppException
import io.realm.mongodb.Credentials
import io.realm.mongodb.ErrorCode
import io.realm.mongodb.User
import io.realm.mongodb.close
import io.realm.mongodb.registerUserAndLogin
import io.realm.rule.BlockingLooperThread
import org.bson.types.ObjectId
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Ignore
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.test.assertEquals

/**
 * Remember: Any test against the local BAAS server will not use an SSL connection.
 */
@RunWith(AndroidJUnit4::class)
@Ignore("Should only be run manually")
class SSLConfigurationTests {

    @get:Rule
    val configFactory = TestSyncConfigurationFactory()
    private val looperThread = BlockingLooperThread()
    private lateinit var app: App

    @Before
    fun setUp() {
        app = App("application-0-nfiqi")
    }

    @Test
    fun disableSSLVerification() = looperThread.runBlocking {
        val username = TestHelper.getRandomEmail()
        val password = "password"
        var user: User = app.registerUserAndLogin(username, password)
        var partition = UUID.randomUUID().toString()

        // 1. Copy a valid Realm to the server
        val syncConfig: SyncConfiguration = configFactory.createSyncConfigurationBuilder(user, partition)
                .modules(DefaultSyncSchema())
                .build()
        var realm = Realm.getInstance(syncConfig)
        realm.beginTransaction()
        realm.createObject(SyncDog::class.java, ObjectId()).name = "Foo"
        realm.commitTransaction()

        // make sure the changes gets to the server
        app.sync.getSession(syncConfig).uploadAllLocalChanges()
        realm.close()
        user.logOut()

        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
        // download the uploaded changes.
        user = app.registerUserAndLogin(TestHelper.getRandomEmail(), "password")
        val syncConfigSSL: SyncConfiguration = configFactory.createSyncConfigurationBuilder(user, partition)
                .modules(DefaultSyncSchema())
                .waitForInitialRemoteData()
                .trustedRootCA("trusted_ca.pem") // This is expired and should be ignored
                .disableSSLVerification()
                .build()
        realm = Realm.getInstance(syncConfigSSL)
        val all = realm.where(SyncDog::class.java).findAll()
        try {
            Assert.assertEquals(1, all.size.toLong())
            Assert.assertEquals("Foo", all[0]!!.name)
        } finally {
            realm.close()
        }
        looperThread.testComplete()
    }

    @Test
    fun trustedRootCA_syncShouldFailWithoutTrustedCA() = looperThread.runBlocking {
        val username = TestHelper.getRandomEmail()
        val password = "password"
        var user: User = app.registerUserAndLogin(username, password)

        // 1. Copy a valid Realm to the server
        val syncConfig: SyncConfiguration = configFactory.createSyncConfigurationBuilder(user)
            .modules(DefaultSyncSchema())
            .build()
        var realm = Realm.getInstance(syncConfig)
        realm.beginTransaction()
        realm.createObject(SyncDog::class.java, ObjectId()).name = "Foo"
        realm.commitTransaction()

        // make sure the changes gets to the server
        app.sync.getSession(syncConfig).uploadAllLocalChanges()
        realm.close()
        user.logOut()

        // 2. Local state should now be completely reset. Open the Realm again with a new configuration which should
        // download the uploaded changes.
        user = app.login(Credentials.emailPassword(username, password))
        val syncConfigSSL: SyncConfiguration = configFactory.createSyncConfigurationBuilder(user)
                .name("useSsl")
                .modules(DefaultSyncSchema())
                .trustedRootCA("untrusted_ca.pem")
                .build()
        // waitForInitialRemoteData will throw an Internal error (125): Operation Canceled
        SystemClock.sleep(TimeUnit.SECONDS.toMillis(2))
        realm = Realm.getInstance(syncConfigSSL)
        try {
            realm.syncSession.downloadAllServerChanges()
        } catch (ex: AppException) {
            assertEquals(ErrorCode.CLIENT_SSL_SERVER_CERT_REJECTED, ex.errorCode)
        } finally {
            realm.close()
        }
        looperThread.testComplete()
    }

    @Test
    fun combining_trustedRootCA_and_withoutSSLVerification_willWarn() = looperThread.runBlocking {
        val username = TestHelper.getRandomEmail()
        val password = "password"
        val user: User = app.registerUserAndLogin(username, password)

        val testLogger = TestLogger()
        val originalLevel = RealmLog.getLevel()
        RealmLog.add(testLogger)
        RealmLog.setLevel(LogLevel.WARN)
        configFactory.createSyncConfigurationBuilder(user)
            .name("useSsl")
            .modules(DefaultSyncSchema())
            .trustedRootCA("trusted_ca.pem")
            .disableSSLVerification()
            .build()
        assertEquals(
            "SSL Verification is disabled, the provided server certificate will not be used.",
            testLogger.message
        )
        RealmLog.remove(testLogger)
        RealmLog.setLevel(originalLevel)
        looperThread.testComplete()
    }
}